enumerable.rb

```diff
@@ -0,0 +1,126 @@
+# reimplementing methods
+private
+# Black arcane magic over voodoo
+def validate?(arg, element, equality)
+  if arg.instance_of? Regexp
+    arg.match?(element.to_s)
+  elsif element.instance_of? arg.class
+    arg == element
+  elsif arg.instance_of? Class
+    element.is_a? arg
+  else
+    equality
+  end
+end
+
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    size.times { |i| yield(to_a[i]) }
+    self
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    size.times { |i| yield(to_a[i], i) }
+    self
+  end
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    arr = []
+    my_each { |i| arr << i if yield(i) }
+    arr
+  end
+
+  def my_all?(arg = nil)
+    my_each do |i|
+      condition = block_given? && !yield(i)
+      condition ||= !block_given? && arg.nil? && !i
+      condition ||= !arg.nil? && !validate?(arg, i, false)
+      return false if condition
+    end
+    true
+  end
+
+  def my_any?(arg = nil)
+    my_each do |i|
+      condition = block_given? && yield(i)
+      condition ||= !block_given? && arg.nil? && i
+      condition ||= !arg.nil? && validate?(arg, i, false)
+      return true if condition
+    end
+    false
+  end
+
+  def my_none?(arg = nil)
+    my_each do |i|
+      condition = block_given? && yield(i)
+      condition ||= !block_given? && arg.nil? && i
+      condition ||= !arg.nil? && validate?(arg, i, false)
+      return false if condition
+    end
+    true
+  end
+
+  def my_count(arg = nil)
+    sum = 0
+    my_each do |i|
+      condition = arg.nil? && !block_given?
+      condition ||= block_given? && yield(i)
+      condition ||= !arg.nil? && i == arg
+      sum += 1 if condition
+    end
+    sum
+  end
+
+  def my_map
+    return to_enum(:my_map) unless block_given?
+
+    arr = []
+    my_each { |i| arr << yield(i) }
+    arr
+  end
+
+  def inject_arg_valid?(*arg)
+    acc = arg[0]&.is_a?(Numeric) ? arg[0] : to_a[0]
+    if arg[0]&.is_a? Symbol
+      operation = arg[0]
+    elsif arg[1]&.is_a? Symbol
+      operation = arg[1]
+    end
+    [acc, operation]
+  end
```

Comment:
* [ ] `my_inject` is identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    array = [4,7,8,89,9,6,5,12,4,2,5]
    operation = proc { |sum, n| sum + n  }
    p array.my_inject(&operation) == array.inject(&operation) # false
  ```
* [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(2, :*) == range.inject(2, :*) # false'
  ```

Reply:
Hold my beer..

Reply:
Thank you for your feedback!
Let me fix that issues🍺